common/deadlock/deadlock.go

```diff
@@ -0,0 +1,222 @@
+		ClusterMetadata   cluster.Metadata
+		ShardController   shard.Controller `optional:"true"`
+	}
+
+	config struct {
+		DumpGoroutines    dynamicconfig.BoolPropertyFn
+		FailHealthCheck   dynamicconfig.BoolPropertyFn
+		AbortProcess      dynamicconfig.BoolPropertyFn
+		Interval          dynamicconfig.DurationPropertyFn
+		MaxWorkersPerRoot dynamicconfig.IntPropertyFn
+	}
+
+	deadlockDetector struct {
+		logger       log.Logger
+		healthServer *health.Server
+		config       config
+		roots        []common.Pingable
+		loops        goro.Group
+	}
+
+	loopContext struct {
+		dd      *deadlockDetector
+		root    common.Pingable
+		ch      chan common.PingCheck
+		workers int32
+		ctx     context.Context // cancelled when loop should exit
+	}
+)
+
+func NewDeadlockDetector(params params) *deadlockDetector {
+	roots := []common.Pingable{
+		params.NamespaceRegistry,
+		params.ClusterMetadata,
+	}
+	if params.ShardController != nil {
+		roots = append(roots, params.ShardController)
+	}
+	return &deadlockDetector{
+		logger:       params.Logger,
+		healthServer: params.HealthServer,
+		config: config{
+			DumpGoroutines:    params.Collection.GetBoolProperty(dynamicconfig.DeadlockDumpGoroutines, true),
+			FailHealthCheck:   params.Collection.GetBoolProperty(dynamicconfig.DeadlockFailHealthCheck, true),
+			AbortProcess:      params.Collection.GetBoolProperty(dynamicconfig.DeadlockAbortProcess, false),
+			Interval:          params.Collection.GetDurationProperty(dynamicconfig.DeadlockInterval, 30*time.Second),
+			MaxWorkersPerRoot: params.Collection.GetIntProperty(dynamicconfig.DeadlockMaxWorkersPerRoot, 10),
+		},
+		roots: roots,
+	}
+}
+
+func (dd *deadlockDetector) Start() error {
+	for _, root := range dd.roots {
+		loopCtx := &loopContext{
+			dd:   dd,
+			root: root,
+			ch:   make(chan common.PingCheck),
+		}
+		dd.loops.Go(loopCtx.run)
+	}
+	return nil
+}
+
+func (dd *deadlockDetector) Stop() error {
+	dd.loops.Cancel()
+	// don't wait for workers to exit, they may be blocked
+	return nil
+}
+
+func (dd *deadlockDetector) detected(name string) {
+	dd.logger.Error("deadlock detected", tag.Name(name))
+
+	if dd.config.FailHealthCheck() {
+		dd.logger.Info("marking grpc services unhealthy")
+		dd.healthServer.Shutdown()
+	}
+
+	if dd.config.DumpGoroutines() {
+		dd.dumpGoroutines()
+	}
+
+	if dd.config.AbortProcess() {
+		dd.logger.Fatal("deadlock detected", tag.Name(name))
+	}
+}
+
+func (dd *deadlockDetector) dumpGoroutines() {
+	profile := pprof.Lookup("goroutine")
+	if profile == nil {
+		dd.logger.Error("could not find goroutine profile")
+		return
+	}
+	var b strings.Builder
+	err := profile.WriteTo(&b, 1) // 1 is magic value that means "text format"
+	if err != nil {
+		dd.logger.Error("failed to get goroutine profile", tag.Error(err))
+		return
+	}
+	// write it as a single log line with embedded newlines.
+	// the value starts with "goroutine profile: total ...\n" so it should be clear
+	dd.logger.Info(b.String())
```

Comment:
Do we need some keyword like "deadlock goroutine dump" so it can be searched in log

Reply:
pprof might change the comment line from dump without us knowing it.